W-18068528 add conversation support enhancements mk (#445)
mulesoft/docs-code-builder: modules/ROOT/pages/int-ai-create-integrations.adoc

= Creating Integrations Using Agentforce
:page-deployment-options: cloud-ide, desktop-ide
:page-aliases: int-ai-enable-einstein.adoc

The Agentforce generative flow feature in Anypoint Code Builder transforms your natural language input, which describes your business logic, into a flow that contains the connector operations and processors, in the correct order, to accomplish the specified task. 

For example, if you want to create a flow that reads contacts from Salesforce in JSON format, enter a message describing that task. The message is then transformed into a series of the connector operations and processors needed to execute that task, in this case: Listener, Query, Transformer, and Logger. 

Agentforce generates a flow, which you can then further configure to suit your specific needs. 

[[before-you-begin]]
== Before You Begin

Before you start creating your integration:

* xref:start-acb.adoc[Set up and access the web or desktop IDE].
* Ensure that you have the xref:start-configure-permissions.adoc#permissions[required Anypoint Code Builder permissions].

> * Ensure that you have the following permissions to use the Anypoint Code Builder Agentforce feature:

Reviewer: ```suggestion
* Ensure that you have these permissions to use the Anypoint Code Builder Agentforce feature:
```

** Anypoint Code Builder Developer
** Mule Developer Generative AI User
* Einstein must be enabled in Access Management. For more information, see xref:access-management::enabling-einstein.adoc[Enabling Einstein for Anypoint Platform].

[[access-einstein]]
== Access Agentforce

There are two ways to open Agentforce:

* When you create a new integration project, click a card in the flow canvas. For example, the *Build a Flow with Agentforce* card.
* Click the Agentforce icon in the toolbar. 
+

> image::int-einstein-icon-in-toolbar.png[Arrow pointing to Agentforce icon in the toolbar]

Reviewer: ```suggestion
image::int-einstein-icon-in-toolbar.png["Arrow pointing to Agentforce icon in the toolbar"]
```

* Specify the data objects.
+
Clearly mention the data objects to include in the flow, such as "Salesforce Account object" or any other relevant data source.
* Proofread your message.
+
Review your message to ensure that it's clear and comprehensive before you send it.

After Agentforce generates the initial flow, you can then customize the code for your specific needs. Or, you can continue the conversation to have Agentforce revise the flow after each message you send.

For examples of messages and generated flows, see xref:int-ai-gen-flow-examples.adoc[].

[NOTE]
====
* You can send up to 20 messages per chat. If you reach the limit, start a new chat.

> * Due to the sensitivity of LLMs and depending on which LLM is used, the same message might produce different responses. Sometimes, a similar message can produce very different responses. 

Reviewer: ```suggestion
* Due to the sensitivity of LLMs and depending on which LLM is used, the same message can produce different responses. Sometimes, a similar message can produce different responses. 
```

====

[[view-prompt-history]]
== View Your Chat History

The message history for a project shows the previous 50 messages for that project. The history can be helpful for reference during the development process.

To see the chat history:

. Click the open chat history icon.
+
You see a list of your previous messages, up to the maximum of 50. You can rename your chats in chat history.
. Click a message in the list to see the details of the generated flow. 
+
You can insert the code that was generated into your flow, or you can copy the code snippet and use or modify it elsewhere. 

[[create-gen-flow]]
== Add a Flow to Your Project

Add a flow to your integration project using Agentforce:

. In the *Build a Flow with Agentforce* card, click *Use Agentforce*.
+
Optionally, click an example on the *Build a Flow with Agentforce* card to see how this feature works.

> . If this is the first time you're using this feature, agree to the terms, and continue. 

Reviewer: ```suggestion
. If this is the first time you use this feature, agree to the terms, and continue. 
```

Reviewer: Are only "terms" or "terms and conditions"?

Author: just terms

+
The *Agentforce* panel opens:
+
image::acb-einstein-panel.png["Agentforce panel with welcome message and examples"]
+
. Enter a message to create the initial flow for your task.
+
Optionally, click an example to get started.
+
When your flow is ready, the code and a description of the components in the flow appear in the *Agentforce* panel:
+
image::int-generated-case-notification-email.png[]
+
. Optionally, click the thumbs up or thumbs down icon that appears after the generated code. 
+
Rating the flow that's created for you helps improve the quality of generative flows.

> . Inspect the generated code, and if you're satisfied, click the insert code icon to insert the code into your flow. 

Reviewer: ```suggestion
. Inspect the generated code. If you're satisfied, click the insert code icon to add the code to your flow. 
```

+
If you aren't satisfied with the generated flow, try rewording your message and sending it again. 

[[customize-gen-flow]]
== Complete Your Flow

> After Agentforce builds the flow, you might need to fill in information specific to your configuration. 

Reviewer: ```suggestion
After Agentforce builds the flow, fill in information specific to your configuration as needed. 
```
